### PR TITLE
add multiprocessing

### PR DESCRIPTION
**Pull request by LFT18**

### CODE REVIEW

File: src/move/data/perturbations.py [hunk 1]
```diff
@@ -149,22 +352,27 @@ def perturb_continuous_data_extended(
...
     assert baseline_dataset.con_shapes is not None
     assert baseline_dataset.con_all is not None
 
+    logger.debug("Creating target_ics, splits, and slice")
     target_idx = con_dataset_names.index(target_dataset_name)  # dataset index
     splits = np.cumsum([0] + baseline_dataset.con_shapes)
     slice_ = slice(*splits[target_idx : target_idx + 2])
 
-    num_features = baseline_dataset.con_shapes[target_idx]
+    #num_features = baseline_dataset.con_shapes[target_idx]
```
Comment (enryH) [inline src/move/data/perturbations.py:365]: Is this correct? Or does it still need to be changed?

Comment (LFT18) [reply src/move/data/perturbations.py:365]: I think it's correct now. I cannot check right now because I've been having problems connecting to Esrum all morning, but I'll check as soon as it works again (hopefully soon)

Comment (LFT18) [reply src/move/data/perturbations.py:365]: I still can't connect, it's very annoying :( . I'll let you know as soon as I can again, but I think the code should be fine

File: analysis/__init__.py
Comment (enryH) [inline analysis/__init__.py:1]: Reminder to me: Check folder and most likely delete entire folder as duplicate.

File: isoforms_first_try/config/data/isoforms.yaml
Comment (enryH) [inline isoforms_first_try/config/data/isoforms.yaml:1]: I will create an extra branch with your analysis scripts/configs - and then merge the core functionality to the developer branch.

File: src/move/__main__.py [hunk 2]
```diff
@@ -4,6 +4,12 @@
 from omegaconf import OmegaConf
 
 import move.tasks
+import move.tasks.analyze_latent_fast
+import move.tasks.analyze_latent_multiprocess
+import move.tasks.identify_associations_efficient
```
Comment (enryH) [inline src/move/__main__.py:9]: intermediate (can be removed)

File: src/move/tasks/analyze_latent_fast.py
Comment (enryH) [inline src/move/tasks/analyze_latent_fast.py:1]: can be removed, just does some of the latent analysis ( subset of task)

File: src/move/data/perturbations.py [hunk 3]
```diff
@@ -1,4 +1,9 @@
-__all__ = ["perturb_categorical_data", "perturb_continuous_data"]
+__all__ = [
+    "perturb_categorical_data",
+    "perturb_continuous_data",
+    "perturb_continuous_data_extended_one",
```
Comment (enryH) [inline src/move/data/perturbations.py:4]: check if this can be the new, memory-efficient default

File: src/move/data/perturbations.py [hunk 4]
```diff
@@ -7,13 +12,229 @@
...
+    perturbed_dataloader = DataLoader(
+        perturbed_dataset,
+        shuffle=False,
+        batch_size=baseline_dataloader.batch_size,
+    )
+
+    return perturbed_dataloader
+
+
+def perturb_continuous_data_extended_one(  # We will keep the input almost the same, to make everything easier
```
Comment (enryH) [inline src/move/data/perturbations.py:131]: adapted to return only one dataloader (from Marc's adaptions)

File: src/move/tasks/identify_associations.py [hunk 5]
```diff
@@ -276,26 +513,80 @@ def _bayes_approach(
...
+    # )  # Bring feature_i feature_i associations to minimum
     sort_ids = np.argsort(bayes_abs, axis=None)[::-1]  # 1D: N x C
+    file_path = output_path / "sort_ids_original_script.tsv"
+    logger.debug(f"Saving sort_ids to {file_path}")
+    np.savetxt(file_path, sort_ids, delimiter="\t")
+    logger.debug(f"sort_ids are {sort_ids}")
+
     prob = np.take(bayes_p, sort_ids)  # 1D: N x C
+    file_path = output_path / "prob_original_final.tsv"
+    logger.debug(f"Saving prob to {file_path}")
```
Comment (enryH) [inline src/move/tasks/identify_associations.py:570]: discuss if all the additionally saved files are needed for the MOVE workflows

File: src/move/conf/schema.py [hunk 6]
```diff
@@ -134,6 +134,9 @@ class AnalyzeLatentConfig(TaskConfig):
 
     feature_names: list[str] = field(default_factory=list)
     reducer: dict[str, Any] = MISSING
+    fast: bool = False  # Default value is False
```
Comment (enryH) [inline src/move/conf/schema.py:137]: this can be deleted as well

File: src/move/data/preprocessing.py [hunk 7]
```diff
@@ -74,7 +74,9 @@ def scale(x: np.ndarray) -> tuple[FloatArray, BoolArray]:
         Tuple containing (1) scaled output and (2) a 1D mask marking columns
         (i.e., features) without zero variance
     """
-    logx = np.log2(x + 1)
+    # CHANGES: Allow for negative values, so only standardization
```
Comment (enryH) [inline src/move/data/preprocessing.py:77]: What was the reason for that change?

**Comment by enryH** [comment]
@ri-heme, @LFT18 ,

so far we do not get the same results for the same configuration of the small dataset in the CI. I checked now everything and I cannot see what it is. 

I think  the last step is to change `identify_associations` the same way as `analyze_latent`: the single process should also only use one perturbed dataloader at the time.

**Comment by LFT18** [comment]
Hi! I can stop by tomorrow or on Wednesday and we can check if there is anything wrong with my code, @enryH

**Comment by enryH** [comment]
> Hi! I can stop by tomorrow or on Wednesday and we can check if there is anything wrong with my code, @enryH

I am done now with all alignments. Just pass by when it suits you and let me know:)

File: src/move/data/preprocessing.py (hunk 7 above)
Comment (enryH) [reply src/move/data/preprocessing.py:77]: @ri-heme What is your take on negative data? Should the log-normalization be optional? Or should it always be applied?

File: src/move/tasks/identify_associations.py [hunk 8]
```diff
@@ -281,23 +296,53 @@ def _bayes_approach(
...
+
     bayes_abs[bayes_mask] = np.min(
         bayes_abs
     )  # Bring feature_i feature_i associations to minimum
+    # Get only the significant associations:
+    # This will flatten the array, so we get all bayes_abs for all perturbed features
+    # vs all continuous features in one 1D array
+    # Then, we sort them, and get the indexes in the flattened array. So, we get an
+    # list of sorted indexes in the flatenned array
     sort_ids = np.argsort(bayes_abs, axis=None)[::-1]  # 1D: N x C
```
Comment (enryH) [inline src/move/tasks/identify_associations.py:311]: So the flattening here means that the probabilities and FDR are calculated across perturbations (meaning that actually pertubations increase the number of total probabilities)?

**Comment by enryH** [comment]
To solve: Reloading trained models from single process with single process yields not exactly the same `fdr` array everytime. The multiprocess version in general higher `fdr` values -> we need to figure out why

**Comment by enryH** [comment]
I finally found the issue. The multiprocessing did no yet have categorical vs continous pertubations implemented. I moved the masking into the main function and update the bayes_worker fct accordingly. I think it's ready to be checked now @ri-heme

File: src/move/tasks/identify_associations.py [hunk 9]
```diff
@@ -175,15 +163,9 @@ def prepare_for_continuous_perturbation(
...
+    logger.debug(f"Dataloaders length is {len(dataloaders)}")
 
-    assert baseline_dataset.con_all is not None
-    orig_con = baseline_dataset.con_all
-    nan_mask = (orig_con == 0).numpy()  # NaN values encoded as 0s
-    logger.debug(f"# NaN values: {np.sum(nan_mask)}/{orig_con.numel()}")
-    feature_mask = nan_mask
-
-    return (dataloaders, nan_mask, feature_mask)
+    return dataloaders
```
Comment (ri-heme) [inline src/move/tasks/identify_associations.py:168]: Return changed, but not the type hint of this function

Comment (ri-heme) [reply src/move/tasks/identify_associations.py:168]: Same for `prepare_for_categorical_perturbation`

File: src/move/tasks/analyze_latent.py [hunk 10]
```diff
@@ -252,20 +397,80 @@ def analyze_latent(config: MOVEConfig) -> None:
...
+    # NOW, THE SAME BUT FOR CONTINUOUS DATA
 
     for i, dataset_name in enumerate(config.data.continuous_names):
         logger.debug(f"Generating plot: feature importance '{dataset_name}'")
-        dataloaders = perturb_continuous_data(
-            test_dataloader, config.data.continuous_names, dataset_name, 0.0
-        )
-        num_features = len(dataloaders)
+        # NOT SURE IF IT WORKS THE SAME FOR CONTINUOUS FEATURES, CHECK THIS
+        # I did sth that did not work, I'll try again now.
```
Comment (ri-heme) [inline src/move/tasks/analyze_latent.py:410]: Not sure about leaving these comments. They would work better as GitHub issues.

I'm not even sure how to interpret them.

File: src/move/tasks/identify_associations.py [hunk 11]
```diff
@@ -9,6 +9,7 @@
 import numpy as np
 import pandas as pd
 import torch
+import torch.multiprocessing
```
Comment (ri-heme) [inline src/move/tasks/identify_associations.py:12]: Unused import if I'm not mistaken

File: src/move/conf/schema.py [hunk 12]
```diff
@@ -170,6 +172,8 @@ class IdentifyAssociationsConfig(TaskConfig):
 class IdentifyAssociationsBayesConfig(IdentifyAssociationsConfig):
     """Configure the probabilistic approach to identify associations."""
 
+    multiprocess: bool = False  # Default value is False
```
Comment (ri-heme) [inline src/move/conf/schema.py:175]: Would make more sense to add this as an attribute of the base class (`IdentifyAssociationsConfig`), since it will be shared among the child classes (Bayes, Ttest, Ks).

**Review by ri-heme — COMMENTED** [review]
Hi Henry. Left some comments here and there (some type hints were changed, some comments are still there, and the some questions/suggestions).

Thanks for your help!

File: .gitignore
[the remark this answers is not in the document]
Comment (enryH) [reply .gitignore:1]: I will checkout the developer version

File: src/move/tasks/analyze_latent.py (hunk 10 above)
Comment (enryH) [reply src/move/tasks/analyze_latent.py:410]: I delete it

File: src/move/tasks/encode_data.py
[the remark this answers is not in the document]
Comment (enryH) [reply src/move/tasks/encode_data.py:1]: Nice catch. Indeed this is probably an issue which happend due to merge conflicts at some point. I updated it.

File: src/move/tasks/identify_associations.py [hunk 13]
```diff
@@ -236,32 +219,38 @@ def _bayes_approach(
...
-            baseline_recon, axis=0
+        # For each model j, we get a different reconstruction for the baseline.
+        # We haven't perturbed anything yet, we are just
+        # getting the reconstruction for the baseline, to make sure that we get
+        # the same reconstruction for each refit, we cannot
+        # do it inside each process because the results might be different
+        reconstruction_path = (
+            models_path / f"baseline_recon_{task_config.model.num_latent}_{j}.pt"
         )
+        if reconstruction_path.exists():
```
[the remark this answers is not in the document]
Comment (enryH) [reply src/move/tasks/identify_associations.py:237]: Yes, but like this it is easier to compare `bayes_approach` with `bayes_parallel`

File: src/move/tasks/identify_associations.py [hunk 14]
```diff
@@ -270,7 +259,8 @@ def _bayes_approach(
         mask = feature_mask[:, [i]] | nan_mask  # 2D: N x C
         diff = np.ma.masked_array(mean_diff[i, :, :], mask=mask)  # 2D: N x C
         prob = np.ma.compressed(np.mean(diff > 1e-8, axis=0))  # 1D: C
-        bayes_k[i, :] = np.log(prob + 1e-8) - np.log(1 - prob + 1e-8)
+        computed_bayes_k = np.log(prob + 1e-8) - np.log(1 - prob + 1e-8)
```
[the remark this answers is not in the document]
Comment (enryH) [reply src/move/tasks/identify_associations.py:262]: to show where a worker function could be called.

File: src/move/tasks/identify_associations.py (hunk 9 above)
Comment (enryH) [reply src/move/tasks/identify_associations.py:168]: true... I'll fix that

File: src/move/tasks/analyze_latent.py (hunk 10 above)
Comment (enryH) [reply src/move/tasks/analyze_latent.py]: deleted

File: src/move/tasks/identify_associations.py (hunk 11 above)
Comment (enryH) [reply src/move/tasks/identify_associations.py:12]: deleted

File: src/move/conf/schema.py (hunk 12 above)
Comment (enryH) [reply src/move/conf/schema.py]: true. I added it to the base class

**Comment by enryH** [comment]
@ri-heme just merge if you thinks it's fine now:)

File: src/move/tasks/identify_associations.py (hunk 8 above)
Comment (mpielies) [reply src/move/tasks/identify_associations.py]: I think so, yes.
